x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java

```diff
@@ -665,17 +663,94 @@ private void onFailure(Exception e) {
         }
     }
 
+    private void runComputeOnDataNode(
+        CancellableTask task,
+        String externalId,
+        PhysicalPlan reducePlan,
+        DataNodeRequest request,
+        ActionListener<ComputeResponse> listener
+    ) {
+        final List<DriverProfile> collectedProfiles = request.configuration().profile()
+            ? Collections.synchronizedList(new ArrayList<>())
+            : List.of();
+        final var responseHeadersCollector = new ResponseHeadersCollector(transportService.getThreadPool().getThreadContext());
+        listener = ActionListener.runBefore(listener, responseHeadersCollector::finish);
+        try (RefCountingListener refs = new RefCountingListener(listener.map(i -> new ComputeResponse(collectedProfiles)))) {
+            final AtomicBoolean cancelled = new AtomicBoolean();
+            // run compute with target shards
+            var internalSink = exchangeService.createSinkHandler(request.sessionId(), request.pragmas().exchangeBufferSize());
+            DataNodeRequestExecutor dataNodeRequestExecutor = new DataNodeRequestExecutor(
+                request,
+                task,
+                internalSink,
+                request.configuration().pragmas().maxConcurrentShardsPerNode(),
+                collectedProfiles,
+                ActionListener.runBefore(cancelOnFailure(task, cancelled, refs.acquire()), responseHeadersCollector::collect)
+            );
+            dataNodeRequestExecutor.start();
+            // run the node-level reduction
+            var externalSink = exchangeService.getSinkHandler(externalId);
+            var exchangeSource = new ExchangeSourceHandler(1, esqlExecutor);
+            exchangeSource.addRemoteSink(internalSink::fetchPageAsync, 1);
```

Comment:
I think it's confusing that we call the same method `addRemoteSink` for any kind of sink (remote/local). I know that we do not differentiate between them when it comes to functionality, but following the code is much harder like this. Maybe RemoteSink is a misnomer.

The ExchangeService is the one dealing with Sinks and Sources but it also has some static methods that receive a TransportService (implying the "remote" functionality). But, `ExchangeSinkHandler` is the one that creates a `LocalExchangeSink`, while the `ExchangeSourceHandler` can add a RemoteSink and create a `LocalExchangeSource`.
What I'm trying to point out is that it is difficult to establish what is "remote" and what is "local", who is in charge of managing these two types of sides.

Reply:
The RemoteSink doesn't necessarily have to be network . I opened https://github.com/elastic/elasticsearch/pull/106245 to rename `LocalExchangeSink` and `LocalExchangeSink`. These should not have the term `Local`. I will introduce an actual LocalExchange to reduce the number of buffers, and this should be no longer a concern.